File: 3DS/03/source/two.cpp
#include <stdlib.h>
#include <3ds.h>
#include <iostream>
#include <fstream>
#include <vector>
#include <set>
#include <string>
#include "utils.h"

using namespace std;

int two()
{
	auto diagnosis = readlines("romfs:/input.txt");

	int bitlen = diagnosis[0].length();
	set<string> candidates(diagnosis.begin(), diagnosis.end());

	int oxygen = 0;
	for (int i = 0; i < bitlen; i++)
	{
		int bit = 0;
		for (auto line : candidates)
		{
			bit += line[i] == '1' ? 1 : -1;
		}
		int mcb = bit < 0 ? 0 : 1;
		oxygen <<= 1;
		oxygen |= mcb;

		set<string> candidates_cpy(candidates);
		for (auto candidate : candidates_cpy)
		{
			if (candidate[i] != mcb + '0')
			{
				candidates.erase(candidate);
			}
		}
	}

	int CO2 = 0;
	candidates = set<string>(diagnosis.begin(), diagnosis.end());

	for (int i = 0; i < bitlen; i++)
	{
		int bit = 0;
		for (auto line : candidates)
		{
			bit += line[i] == '1' ? 1 : -1;
		}
		int lcb = bit < 0 ? 1 : 0;
		CO2 <<= 1;
		CO2 |= lcb;

		set<string> candidates_cpy(candidates);
		for (auto candidate : candidates_cpy)
		{
			if (candidate[i] != lcb + '0')
			{
				candidates.erase(candidate);

				if (candidates.size() == 1)
				{
					for (int j = i + 1; j < bitlen; j++)
					{
						CO2 <<= 1;
						CO2 |= ((*candidates.begin())[j] - '0');
					}

					return oxygen * CO2;
				}
			}
		}
	}

	return -1;
}

File: 03/two.lua
package.path = package.path .. ";../modules/?.lua"
local tools = require "tools"

diagnosis = tools.readlines("input.txt")

bit_len = #diagnosis[1]

candidates = {}
for _, line in ipairs(diagnosis) do
	candidates[line] = true
end

oxygen = 0

for i = 1, bit_len do
	bits = {[0] = 0, [1] = 0}
	len_candidates = 0
	for candidate, valid in pairs(candidates) do
		bits[tonumber(candidate:sub(i, i))] = bits[tonumber(candidate:sub(i, i))]- -1
		len_candidates = len_candidates- -1
	end

	mcb = bits[0] > bits[1] and 0 or 1
	oxygen = oxygen << 1 | mcb

	for candidate, valid in pairs(candidates) do
		if candidate:sub(i, i) ~= tostring(mcb) then
			candidates[candidate] = nil
		end
	end
end

for _, line in ipairs(diagnosis) do
	candidates[line] = true
end

CO2 = 0
for i = 1, bit_len do
	bits = {[0] = 0, [1] = 0}
	for candidate, valid in pairs(candidates) do
		bits[tonumber(candidate:sub(i, i))] = bits[tonumber(candidate:sub(i, i))]- -1
	end

	lcb = bits[0] <= bits[1] and 0 or 1
	CO2 = CO2 << 1 | lcb

	len_candidates = 0
	for candidate, valid in pairs(candidates) do
		if candidate:sub(i, i) ~= tostring(lcb) then
			candidates[candidate] = nil
		else
			len_candidates = len_candidates- -1
		end
	end

	if len_candidates == 1 then
		for j = i + 1, bit_len do
			CO2 = CO2 << 1 | tonumber(next(candidates, nil):sub(j, j))
		end
		break
	end
end

print(oxygen, CO2)

print("Oxygen generator rating * CO2 scrubber rating = " .. oxygen * CO2)
assert(oxygen * CO2 == 7440311)
